### PR TITLE
kernel: thread stack related cleanups

### PR DESCRIPTION
**Pull request by tyrantlizard77**
This PR:

* Reduces z_new_thread_init() to bare essentials. I'd like to eventually remove this, but that is pending some additional work.
* Centralize some thread priority assertions to common code
* Thread stack macros moved out of kernel.h into to their own header
* New arch interface macro, ARCH_STACK_PTR_ALIGN, replacing what was inconsistently defined as STACK_ALIGN or STACK_ALIGN_SIZE
* New common interface Z_STACK_PTR_ALIGN(size), replacing STACK_ROUND_DOWN()
* Unused STACK_ROUND_UP() deleted

There is a lot of code motion in this PR, it will be simpler to review one patch at a time rather than all at once.

### CODE REVIEW

**Review by nashif — APPROVED** [review]
nice

**Review by dcpleung — APPROVED** [review]

**Review by daor-oti — APPROVED** [review]

**Review by jenmwms — APPROVED** [review]

**Review by andyross — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]
OK for kernel & ARM changes

File: arch/arc/core/thread.c [hunk 1]
```diff
@@ -91,19 +90,19 @@ void arch_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
...
 
-		stackAdjEnd = (char *)STACK_ROUND_DOWN(stackEnd +
+		stackAdjEnd = (char *)Z_STACK_PTR_ALIGN(stackEnd +
 				ARCH_THREAD_STACK_RESERVED);
 
 		/* reserve 4 bytes for the start of user sp */
 		stackAdjEnd -= 4;
-		(*(u32_t *)stackAdjEnd) = STACK_ROUND_DOWN(
+		(*(u32_t *)stackAdjEnd) = Z_STACK_PTR_ALIGN(
 			(u32_t)stackEnd - offset);
```
Comment (stephanosio) [inline arch/arc/core/thread.c:99]: This is somewhat unrelated to this PR, but I wonder if we should fix this while we are at it (AFAICS, we do not need to do pointer-to-integer conversion here at all).

File: kernel/include/kernel_internal.h [hunk 2]
```diff
@@ -46,9 +46,21 @@ extern void z_setup_new_thread(struct k_thread *new_thread,
...
+#if !defined(CONFIG_INIT_STACKS) && !defined(CONFIG_THREAD_STACK_INFO)
+	ARG_UNUSED(stack);
+	ARG_UNUSED(stack_size);
+#endif
+#ifdef CONFIG_INIT_STACKS
+	memset(stack, 0xaa, stack_size);
+#endif
+#if defined(CONFIG_THREAD_STACK_INFO)
+	thread->stack_info.start = (uintptr_t)stack;
+	thread->stack_info.size = (u32_t)stack_size;
```
Comment (stephanosio) [inline kernel/include/kernel_internal.h:61]: This should be `thread->stack_info.size = (size_t)stack_size;`

**Review by stephanosio — COMMENTED** [review]
I see that the `Z_STACK_PTR_ALIGN` currently relies on the `ROUND_DOWN` macro, and this macro casts the input value pointers to `unsigned long`, which is not so 64-bit friendly (we should be using `uintptr_t` for this purpose):

https://github.com/zephyrproject-rtos/zephyr/blob/56aac64bae7b48c13442262b7c21e1a443fd4f9a/include/sys/util.h#L84-L85

Given that the `ROUND_UP` and `ROUND_DOWN` macros are intended to be generic (i.e. not specifically for pointers), I wonder if we should define an additional set of rounding macros for pointers (e.g. `ROUND_UP_PTR` and `ROUND_DOWN_PTR`).

File: kernel/include/kernel_internal.h (hunk 2 above)
Comment (andrewboie) [reply kernel/include/kernel_internal.h:61]: fixed.

**Comment by andrewboie** [comment]
> input value pointers to unsigned long, which is not so 64-bit friendly (we should be using uintptr_t for this purpose):

unsigned long will hold a pointer value on every arch I can think of. I'd like to not address this in this PR, this has been this way for a very long time.

**Pull request merged**